### ISSUE TITLE
GEL PanelApp table only shows partial data (evidence page)

### ISSUE THREAD

**Issue by LucaFumis** [opentargets/issues#2750]
**Describe the bug**
On the evidence page, the **GEL PanelApp** widget doesn't show all data: the widget summary (at top of page) reports a number of entries (41 in the example below), but the actual table only displays 25. 
**Gene2Phenotype** might be affected as well, but need an appropriate example to confirm.
See the _Additional Context_ section below.

**Observed behaviour**
In cases where the GEL PanelApp reports more than 25 entries, only 25 will be shown. Other examples would look correct.

**Expected behaviour**
The GEL PanelApp table should display all the data (i.e. 41 rows in the example below)

**To Reproduce**
Steps to reproduce the behaviour:
1. Go to https://platform.opentargets.org/evidence/ENSG00000171862/MONDO_0017623
2. See summary for GEL PanelApp (near top of page): it shows **41 entries**
3. Click on it to scroll down to the widget table
4. See table only displays 25 rows

**Additional context**
This is due to a front-end / API issue. I've come across it while investigating evidence widgets queries in relation to the new associations table.
Evidence widgets works like this:

1. summary widget gets the total count for entries, n
2. such count n is passed to the query for the widget
3. the API returns n rows

We're currently updating the queries, removing the `size` parameter.
My understanding from various conversations was that the API would return _everything_ when not specifying the size.
However it seems that the API by default returns 25 rows max.

Now back to the current bug.
It seems that **GEL PanelApp** and **Gene2Phenotype** widgets do **not** specify a size in the query. Never did.
So was the API, at some point, returning all data for those widgets? 
Or perhaps they never have more than 25 rows?

**Comment by LucaFumis** [comment]
<img width="1231" alt="panelapp-summary" src="https://user-images.githubusercontent.com/8592363/195654824-baf0d150-5df4-47de-ba8c-0eeca1dfbf25.png">

<img width="1233" alt="panelapp-widget" src="https://user-images.githubusercontent.com/8592363/195654840-ad915c82-ca00-4979-b0fd-cffbd1e4ce29.png">

**Comment by d0choa** [comment]
Hi @LucaFumis, I think this bug must have been there for quite a while, but because PanelApp and gene2phenotype very rarely go beyond 25 records, we never noticed.

The API behaviour is the expected one. Return 25 rows by default to protect the BE from random queries. We handle this nicely in other tables of the evidence page (e.g. genetics portal).

Can you make it work similarly for production purposes?

Regarding the return of all records without knowing the size of the response beforehand. We might need to bring @opentargets/be-team and @opentargets/fe-team to the discussion. I visualise 3 options:
1. query with `size=99999`
2. return all records when specifying `size=-1`, by implementing it in the API
3. query twice the API, once to get the size of the response and another one to get the records with the specified `size`
Please share your thoughts

**Comment by LucaFumis** [comment]
Thanks for the info @d0choa - great feedback.
It makes sense that those widgets rarely go over 25 rows - the queries for PanelApp and gene2phenotype have indeed been like this for a long time (initial rewrite). I guess it's not so critical then,

> Can you make it work similarly for production purposes?
To fix the issue in production we could update the two queries with the correct size (like we do for all other widgets).

> Regarding the return of all records without knowing the size of the response beforehand
Yes, definitely need BE and FE teams ! I was thinking exactly the same options:
> query with size=99999
> return all records when specifying size=-1, by implementing it in the API
My thought as well: fixed big size only needs FE changes. The `-1` option also makes sense to me.
> query twice the API, once to get the size of the response and another one to get the records with the specified size
The query for the count could be fired only if count not available (i.e. the widget is loaded inside the new associations heatmap), while on the associations page count would be available from the summary (current implementation). Probably also work getting BE thoughts on the two queries.

**Comment by LucaFumis** [comment]
> query with size=99999
> query twice the API, once to get the size of the response and another one to get the records with the specified size

A 4th approach, which came up in an informal discussion with @carcruz, @JarrodBaker and @mbdebian - a mix of the two above: quickly load a small set + count for each widget (so tables are populated quickly), then load the rest. However data would have to be sorted in the API...

**Comment by carcruz** [comment]
Just validate with the current associations on the fly POC and this is. a real bug that is affecting all evidence API responses. API only returns 25 records. What do you think will be a good approach to follow @JarrodBaker?

**Comment by JarrodBaker** [comment]
Can you please provide the query you're using @carcruz  so we make sure we're all looking at the same thing?  Thx

**Comment by carcruz** [comment]
Sure, this query is showing `count: 27` but only returning 25 records without the `size`
```
query GenomicsEnglandQuery {
  disease(efoId: "EFO_0000305") {
    id
    evidences(
      ensemblIds: ["ENSG00000139618"]
      enableIndirect: true
      datasourceIds: ["genomics_england"]
    ) {
      count
      rows {
        disease {
          id
          name
        }
        target {
          approvedSymbol
        }
        diseaseFromSource
        cohortPhenotypes
        confidence
        allelicRequirements
        studyOverview
        studyId
        confidence
        literature
      }
    }
  }
}

```

**Comment by JarrodBaker** [comment]
I might be missing something to understand the problem correctly, but I think the API has the fields that are needed already. 

On the API there are parameters `size` and `cursor` on `Evidence`, so we can update Carlos' query with: 

```
query GenomicsEnglandQuery {
  disease(efoId: "EFO_0000305") {
    id
    evidences(
      ensemblIds: ["ENSG00000139618"]
      enableIndirect: true
      datasourceIds: ["genomics_england"]
      size:50
    ) {
      count <-- total records available
      cursor <-- link to next page of results
      rows {
        disease {
          id
          name
        }
        target {
          approvedSymbol
        }
        diseaseFromSource
        cohortPhenotypes
        confidence
        allelicRequirements
        studyOverview
        studyId
        confidence
        literature
      }
    }
  }
}
```
Note I've added in the `size` query parameter, and also requested that the `cursor` field be returned. 

The response is:
```
{
  "data": {
    "disease": {
      "id": "EFO_0000305",
      "evidences": {
        "count": 27,
        "cursor": "WzEsIjJlMGUyNTcyM2NkYzA5YmM5OWYzMGZhM2YzMzJkYWRjNmNhNWEyNGQiXQ==",
        "rows": [
          {
            "disease": {
              "id": "Orphanet_227535",
              "name": "Hereditary breast cancer"
            },
            "target": {
              "approvedSymbol": "BRCA2"
            },
            "diseaseFromSource": "Breast cancer, male, susceptibility to",
            "cohortPhenotypes": [
              "Breast-ovarian cancer, familial, 2, OMIM:612555",
              "Breast cancer, male, susceptibility to, OMIM:114480"
            ],
            "confidence": "green",
            "allelicRequirements": [
              "MONOALLELIC, autosomal or pseudoautosomal, NOT imprinted"
            ],
            "studyOverview": "Familial breast cancer",
            "studyId": "158",
            "literature": null
          },
          {
            "disease": {
              "id": "Orphanet_227535",
              "name": "Hereditary breast cancer"
            },
            "target": {
              "approvedSymbol": "BRCA2"
            },
            "diseaseFromSource": "Breast cancer, male, susceptibility to",
            "cohortPhenotypes": [
              "Breast-ovarian cancer, familial, 2, OMIM:612555",
              "Breast cancer, male, susceptibility to, OMIM:114480",
              "Adult only"
            ],
            "confidence": "green",
            "allelicRequirements": [
              "MONOALLELIC, autosomal or pseudoautosomal, imprinted status unknown"
            ],
            "studyOverview": "Additional findings health related - adult specific",
            "studyId": "929",
            "literature": null
          },
          {
            "disease": {
              "id": "MONDO_0016419",
              "name": "hereditary breast carcinoma"
            },
            "target": {
              "approvedSymbol": "BRCA2"
            },
            "diseaseFromSource": "Breast cancer, male, susceptibility to",
            "cohortPhenotypes": [
              "Breast-ovarian cancer, familial, 2, OMIM:612555",
              "Breast cancer, male, susceptibility to, OMIM:114480",
              "Prostate cancer, OMIM:176807"
            ],
            "confidence": "green",
            "allelicRequirements": [
              "MONOALLELIC, autosomal or pseudoautosomal, imprinted status unknown"
            ],
            "studyOverview": "Adult solid tumours cancer susceptibility",
            "studyId": "245",
            "literature": null
          },
          {
            "disease": {
              "id": "MONDO_0012933",
              "name": "breast-ovarian cancer, familial, susceptibility to, 2"
            },
            "target": {
              "approvedSymbol": "BRCA2"
            },
            "diseaseFromSource": "Breast-ovarian cancer, familial, 2",
            "cohortPhenotypes": [
              "Breast-ovarian cancer, familial, 2, OMIM:612555",
              "Breast cancer, male, susceptibility to, OMIM:114480"
            ],
            "confidence": "green",
            "allelicRequirements": [
              "MONOALLELIC, autosomal or pseudoautosomal, NOT imprinted"
            ],
            "studyOverview": "Familial breast cancer",
            "studyId": "158",
            "literature": null
          },
          {
            "disease": {
              "id": "EFO_0000305",
              "name": "breast carcinoma"
            },
            "target": {
              "approvedSymbol": "BRCA2"
            },
            "diseaseFromSource": "Breast cancer, male, susceptibility to",
            "cohortPhenotypes": [
              "Breast-ovarian cancer, familial, 2, OMIM:612555",
              "Breast cancer, male, susceptibility to, OMIM:114480"
            ],
            "confidence": "green",
            "allelicRequirements": [
              "MONOALLELIC, autosomal or pseudoautosomal, NOT imprinted"
            ],
            "studyOverview": "Familial breast cancer",
            "studyId": "158",
            "literature": null
          }
        ]
      }
    }
  }
}
```
Running the results through `jq` (`jq '.data.disease.evidences.rows | length'`) shows that 5 evidences were returned, as requested. There is also a non-null value returned in the `cursor` field, implying that there are additional records.

Updating the query to include a `cursor` parameter:

```
query GenomicsEnglandQuery {
  disease(efoId: "EFO_0000305") {
    id
    evidences(
      ensemblIds: ["ENSG00000139618"]
      enableIndirect: true
      datasourceIds: ["genomics_england"]
      size:5
      cursor:"WzEsIjJlMGUyNTcyM2NkYzA5YmM5OWYzMGZhM2YzMzJkYWRjNmNhNWEyNGQiXQ=="
    ) {
      count
      cursor
      rows {
        disease {
          id
          name
        }
        target {
          approvedSymbol
        }
        diseaseFromSource
        cohortPhenotypes
        confidence
        allelicRequirements
        studyOverview
        studyId
        confidence
        literature
      }
    }
  }
}
```
returns the next 5 results. If the `cursor` returned from the server is `null`, then there are no more results. 

So to generate the table, we'd use the value from `count` to limit the pagination, update the `size` variable to be the desired page size, and use the `cursor` to get the results as requested. 

Is this the right track?

**Comment by d0choa** [comment]
I think you are.

- We query the API once to populate the first page of the table + get the count of total entries.
- We query the API on user interaction using the cursor for any additional data required (next page, show more rows, etc.)

Are we too far from this @carcruz?

**Comment by LucaFumis** [comment]
That's the approach we were trying with the ClinVar widget which, unlike other widgets, uses the cursor, albeit with a initial size of 1000:
https://github.com/opentargets/ot-ui-apps/blob/lf-query-size/apps/platform/src/sections/evidence/EVA/Body.jsx
https://github.com/opentargets/ot-ui-apps/blob/lf-query-size/apps/platform/src/sections/evidence/EVA/ClinvarQuery.gql
```
query ClinvarQuery(
  $ensemblId: String!
  $efoId: String!
  $size: Int!
  $cursor: String
) {
  disease(efoId: $efoId) {
    id
    evidences(
      ensemblIds: [$ensemblId]
      enableIndirect: true
      datasourceIds: ["eva"]
      size: $size
      cursor: $cursor
    ) {
      cursor
      count
      rows {
        # ... 
      }
    }
  }
}
```

A consistent approach for all widgets would help keeping things clear and easier to maintain.

**Comment by LucaFumis** [comment]
Had a chat about this with @carcruz - the cursor pagination approach looks neat so we'll test that.
@JarrodBaker could I just double-check with you that the `cursor` works for all datasources in the evidence endpoint please?
In the past, if I remember correctly, only a couple of widgets had that server side pagination, leading to different implementations in the front end and bit of confusion.

**Comment by LucaFumis** [comment]
Just a quick note as this is included in the 23.02 release intentions: these issues are being addressed (refactoring evidence widgets) as part of the AOTF project/development.

**Comment by ireneisdoomed** [comment]
This has been solved with [#93](https://github.com/opentargets/ot-ui-apps/pull/93)
